### DISCUSSION TITLE
[Chapter 01] 데이터 클래스

### DISCUSSION THREAD

**Discussion by firewoody237** [cowaine/is-my-code-that-strange#3] [Chapter 01-03 (설계)] [answered]
- 6페이지를 보면 저자는 데이터 클래스를 "악마 생성기"처럼 얘기하고 있습니다.
- `DTO`의 경우 비즈니스 로직이 없고, `VO`와 달리 값을 변경할 수 있게끔 만드는 경우도 있을 것 같은데요, 그러면 저자는 내부 속성의 변경이 필요할 시 `Setter`가 아닌 새로운 객체를 반환해야 한단 걸까요? (저자는 DTO를 싫어할까요..? ㅎㅎ)
- [DTO는 getter와 setter를 포함할 수 있음](https://peonyf.tistory.com/entry/Spring-Entity-DTO-VO)

**Comment by Irisation23** [comment, marked as answer]
데이터 클래스가 악마화 된 이유는 DTO의 개념보다는 예제에 표기된 클래스 `ContractAmount` 가 세금 계산 로직을 구현하지 않아서 생긴 문제라는 생각이 듭니다.

DTO는 계층간 데이터 교환이 이루어지는 것이 목표입니다. 그렇지만 책의 예제는 유지보수를 거치다가 1:N 구조를 가지게 되는 클래스가 발생했고 1을 의존하는 여러 클래스들이 동일 연산을 해야하니, 이는 코드 중복 및 유지 보수에 불편함을 초례 한다고 생각해 볼 수 있습니다.

해당 책의 목차를 좀 읽어보니 불변 객체에 대한 것을 다루는 것으로 미루어 보아 차후에 
```
저자는 내부 속성의 변경이 필요할 시 Setter가 아닌 새로운 객체를 반환해야 한단 걸까요?
```
위의 질문에 해당하는 내용을 담을 것 같습니다!

11p 하부를 참조 해 보시면 DTO는 애초에 설계가 다르다는 관점이 있습니다.

**Reply by Irisation23** [reply]
추가적으로 질문 있습니다.

해당 챕터에 언급되는 `데이터 클래스`의 정의가 public 필드, 기본 생성자만 존재하는 클래스가 맞을까요?

**Reply by 0sunset0** [reply]
> 데이터 클래스는 초기화 작업을 하는 코드조차 다른 클래스에 구현되어 있습니다. 
인스턴스 변수에 어떠한 값이든 넣을 수 있으므로, 잘못된 값이 쉽게 들어간다는 문제도 있습니다. 데이터 클래스를 잘못된 값으로부터 방어하기 위한 유효성 검사도 다른 클래스에 구현되어 있습니다. 즉, 데이터 클래스가 자기 자신을 보호할 수 있는 로직을 갖고 있지 않은 것입니다. -25p

내용을 봤을 때 데이터만 가지고 있고, 외부에서 그 데이터를 조작해야 하므로 
저자가 말하는 데이터 클래스는  public 필드, 기본 생성자만 존재하는 클래스가 맞는 것 같습니다!

**Comment by Irisation23** [comment]
### 해당 파트에서 같이 보면 좋을 내용

- 자바의 `Record`와 코틀린의 `DataClass`는 닮아 있는가??
